core/src/main/java/io/questdb/griffin/engine/join/HashOuterJoinFilteredLightRecordCursorFactory.java

```diff
@@ -217,59 +460,136 @@ public boolean hasNext() {
+                mapCursor = joinKeyMap.getCursor();
+            }
+
+            if (slaveChainCursor != null && slaveChainCursor.hasNext()) {
+                while (slaveChainCursor.hasNext()) {
+                    slaveCursor.recordAt(slaveRecord, slaveChainCursor.next());
+                    if (record.hasMaster()) {
+                        if (filter.getBool(record)) {
+                            MapKey keys = matchIdsMap.withKey();
+                            keys.put(slaveRecord, RecordIdSink.RECORD_ID_SINK);
+                            keys.createValue();
+                            return true;
+                        }
+                    } else {
+                        MapKey keys = matchIdsMap.withKey();
+                        keys.put(slaveRecord, RecordIdSink.RECORD_ID_SINK);
+                        if (keys.findValue() == null) {
+                            return true;
+                        }
+                    }
+                }
+            }
+
+            while (masterCursor.hasNext()) {
                 MapKey key = joinKeyMap.withKey();
-                key.put(record, slaveKeySink);
-                MapValue value = key.createValue();
-                if (value.isNew()) {
-                    final int offset = slaveChain.put(record.getRowId(), -1);
-                    value.putInt(0, offset);
-                } else {
-                    value.putInt(0, slaveChain.put(record.getRowId(), value.getInt(0)));
+                key.put(masterRecord, masterKeySink);
+                MapValue value = key.findValue();
+                if (value != null) {
+                    slaveChainCursor = slaveChain.getCursor(value.getInt(0));
+                    while (slaveChainCursor.hasNext()) {
+                        slaveCursor.recordAt(slaveRecord, slaveChainCursor.next());
+                        if (filter.getBool(record)) {
+                            MapKey keys = matchIdsMap.withKey();
+                            keys.put(slaveRecord, RecordIdSink.RECORD_ID_SINK);
+                            keys.createValue();
+                            return true;
+                        }
+                    }
+                }
+            }
+
+            record.hasMaster(false);
+            while (mapCursor.hasNext()) {
+                MapRecord mapRecord = mapCursor.getRecord();
+                slaveChainCursor = slaveChain.getCursor(mapRecord.getValue().getInt(0));
+                while (slaveChainCursor.hasNext()) {
+                    slaveCursor.recordAt(slaveRecord, slaveChainCursor.next());
+                    MapKey keys = matchIdsMap.withKey();
+                    keys.put(slaveRecord, RecordIdSink.RECORD_ID_SINK);
+                    if (keys.findValue() == null) {
+                        return true;
+                    }
                 }
             }
+            return false;
+        }
+
+        @Override
+        public void toTop() {
```

Comment:
Surprisingly, this method is not covered by `JoinTest`. Could you check the test coverage and change `assertSql` to `assertQueryNoLeakCheck` where appropriate?

Reply:
Fixed to `assertQueryNoLeakCheck` to make tests cover this code.